### PR TITLE
Skip `unnecessary_to_owned` when `t != t.to_string()`

### LINKED ISSUE #9317: unnecessary use of `to_string` for a type with custom `ToString` trait
### Summary

I'm programmatically commiting a patch to a git repo and then writing the hash (a [`git2::Oid`](https://docs.rs/git2/latest/git2/struct.Oid.html)) in a file:

```
let commit = util::commit(...);
fs::write(path, commit);
```

if I use above code, I'll get bytes written to the file whereas I wanted the string of hash, so naturally I change it to `fs::write(path, commit.to_string())`, however that's throwing a lint error:

```
warning: unnecessary use of `to_string`
   --> ...
    |
810 |             commit.to_string(),
    |             ^^^^^^^^^^^^^^^^^^ help: use: `commit`
    |
    = note: `#[warn(clippy::unnecessary_to_owned)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_to_owned
```

solution is to use `format!(...)` instead, but that's pretty ugly.

### Lint Name

unnecessary_to_owned

### Reproducer

I tried this code:

```rust
<code>
```

I saw this happen:

```
<output>
```

I expected to see this happen:

### Version

```text
rustc 1.61.0 (fe5b13d68 2022-05-18)
binary: rustc
commit-hash: fe5b13d681f25ee6474be29d748c65adcd91f69e
commit-date: 2022-05-18
host: x86_64-unknown-linux-gnu
release: 1.61.0
LLVM version: 14.0.0
```

### Additional Labels

_No response_

### PR DESCRIPTION
**Pull request by xphoniex**
Fixes #9317

changelog: [`unnecessary_to_owned`]: none

### CODE REVIEW

File: clippy_lints/src/methods/unnecessary_to_owned.rs [hunk 1]
```diff
@@ -279,6 +280,21 @@ fn check_other_call_arg<'tcx>(
...
+                    if matches!(inner_ty.kind(), ty::Uint(ty::UintTy::U8)) {
+                        let self_ty = cx.typeck_results().expr_ty(expr).peel_refs();
+                        if  is_type_diagnostic_item(cx, self_ty, sym::String) {
+                            return false;
+                        }
+                    }
+                }
+            }
+
             implements_trait(cx, receiver_ty, as_ref_trait_id, &composed_substs)
```
Comment (flip1995) [inline clippy_lints/src/methods/unnecessary_to_owned.rs:298]:
```suggestion
            // if `expr` is a `String` and generic target is [u8], skip
            // (https://github.com/rust-lang/rust-clippy/issues/9317).
            if let [subst] = composed_substs
                && let GenericArgKind::Type(arg_ty) = subst.unpack() 
                && arg_ty.is_slice() 
                && let inner_ty = arg_ty.builtin_index().unwrap()
                && let ty::Uint(ty::UintTy::U8) = inner_ty.kind() 
                && let self_ty = cx.typeck_results().expr_ty(expr).peel_refs()
                && is_type_diagnostic_item(cx, self_ty, sym::String) {
                false
            } else {
                implements_trait(cx, receiver_ty, as_ref_trait_id, &composed_substs)
            }
```

Let's use `let_chain`s here.

**Review by flip1995 — COMMENTED** [review]
Thanks! Small suggestion on how to prettify the code a bit. Other than that, LGTM.

File: clippy_lints/src/methods/unnecessary_to_owned.rs (hunk 1 above)
Comment (xphoniex) [reply clippy_lints/src/methods/unnecessary_to_owned.rs]: https://github.com/rust-lang/rust-clippy/compare/3fa0fba225454cb3aebf5dd973827d6f540de5cf..1a2aaf68de4989eaacf5b669da00b50dca8b41e0

**Comment by xphoniex** [comment]
Great. Applied the changes.

**Comment by flip1995** [comment]
@bors r+

Thanks!

**Pull request merged**